### PR TITLE
optimisation: group requests to filter server

### PR DESCRIPTION
**Pull request by jamal-khey**

### CODE REVIEW

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java [hunk 1]
```diff
@@ -255,6 +262,7 @@ private List<SensitivityVariableSet> buildSensitivityVariableSets(SensitivityAna
                     .build());
                 return;
             }
+
```
Comment (AAJELLAL) [inline src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java:265]: Unnecessary change.

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java [hunk 2]
```diff
@@ -296,9 +304,10 @@ private List<SensitivityFactor> buildSensitivityFactorsFromVariablesSets(Sensiti
             return List.of();
         }
 
-        List<IdentifiableAttributes> monitoredEquipments = monitoredEquipmentsContainers.stream()
-            .flatMap(filter -> getMonitoredIdentifiablesFromContainer(context, network, filter, monitoredEquipmentsTypesAllowed, reporter))
-            .collect(Collectors.toList());
+        // List<IdentifiableAttributes> monitoredEquipments = monitoredEquipmentsContainers.stream()
+        //     .flatMap(filter -> getMonitoredIdentifiablesFromContainer(context, network, filter, monitoredEquipmentsTypesAllowed, reporter))
+        //     .collect(Collectors.toList());
```
Comment (AAJELLAL) [inline src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java:309]: Remove this.

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java [hunk 3]
```diff
@@ -312,17 +321,20 @@ private List<SensitivityFactor> buildSensitivityFactorsFromEquipments(Sensitivit
                                                                           List<Contingency> contingencies,
                                                                           SensitivityFunctionType sensitivityFunctionType,
                                                                           SensitivityVariableType sensitivityVariableType) {
-        List<IdentifiableAttributes> equipments = filters.stream()
-            .flatMap(filter -> getIdentifiablesFromContainer(context, filter, equipmentsTypesAllowed, reporter))
-            .collect(Collectors.toList());
+        // List<IdentifiableAttributes> equipments = filters.stream()
+        //     .flatMap(filter -> getIdentifiablesFromContainer(context, filter, equipmentsTypesAllowed, reporter))
+        //     .collect(Collectors.toList());
```
Comment (AAJELLAL) [inline src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java:326]: To be removed.

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java [hunk 4]
```diff
@@ -312,17 +321,20 @@ private List<SensitivityFactor> buildSensitivityFactorsFromEquipments(Sensitivit
...
         if (equipments.isEmpty()) {
             return List.of();
         }
 
-        List<IdentifiableAttributes> monitoredEquipments = monitoredEquipmentsContainers.stream()
-            .flatMap(filter -> getMonitoredIdentifiablesFromContainer(context, network, filter, monitoredEquipmentsTypesAllowed, reporter))
-            .collect(Collectors.toList());
+        // List<IdentifiableAttributes> monitoredEquipments = monitoredEquipmentsContainers.stream()
+        //     .flatMap(filter -> getMonitoredIdentifiablesFromContainer(context, network, filter, monitoredEquipmentsTypesAllowed, reporter))
+        //     .collect(Collectors.toList());
```
Comment (AAJELLAL) [inline src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java:335]: The same here.

File: src/test/java/org/gridsuite/sensitivityanalysis/server/SensitivityAnalysisControllerTest.java [hunk 5]
```diff
@@ -413,68 +413,70 @@ public void setUp() {
...
+        given(filterService.getIdentifiablesFromFilters(List.of(EQUIPMENTS_IN_VOLTAGE_REGULATION_FILTERS_UUID), NETWORK_UUID, VARIANT_1_ID)).willReturn(EQUIPMENTS_IN_VOLTAGE_REGULATION);
+        given(filterService.getIdentifiablesFromFilters(List.of(EQUIPMENTS_IN_VOLTAGE_REGULATION_FILTERS_UUID), NETWORK_UUID, VARIANT_3_ID)).willReturn(EQUIPMENTS_IN_VOLTAGE_REGULATION_VARIANT);
+        given(filterService.getIdentifiablesFromFilters(List.of(EQUIPMENTS_IN_VOLTAGE_REGULATION_FILTERS_UUID), NETWORK_UUID, VARIANT_2_ID)).willReturn(EQUIPMENTS_IN_VOLTAGE_REGULATION);
+        given(filterService.getIdentifiablesFromFilters(List.of(EQUIPMENTS_IN_VOLTAGE_REGULATION_FILTERS_UUID), NETWORK_UUID, null)).willReturn(EQUIPMENTS_IN_VOLTAGE_REGULATION);
+        given(filterService.getIdentifiablesFromFilters(List.of(EQUIPMENTS_IN_VOLTAGE_REGULATION_FILTERS_UUID), NETWORK_STOP_UUID, VARIANT_2_ID)).willReturn(EQUIPMENTS_IN_VOLTAGE_REGULATION);
         given(filterService.getIdentifiablesCount(IDS_1, NETWORK_UUID, null)).willReturn(Map.<String, List<Long>>of(MONITORED_BRANCHS_KEY, List.<Long>of(6L), INJECTIONS_KEY, List.<Long>of(6L)));
+
```
Comment (AAJELLAL) [inline src/test/java/org/gridsuite/sensitivityanalysis/server/SensitivityAnalysisControllerTest.java:479]: Unnecessary change.

**Review by AAJELLAL — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java (hunk 1 above)
Comment (jamal-khey) [reply src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java]: fixed

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java (hunk 3 above)
Comment (jamal-khey) [reply src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java]: done

File: src/test/java/org/gridsuite/sensitivityanalysis/server/SensitivityAnalysisControllerTest.java (hunk 5 above)
Comment (jamal-khey) [reply src/test/java/org/gridsuite/sensitivityanalysis/server/SensitivityAnalysisControllerTest.java]: done

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java [hunk 6]
```diff
@@ -141,17 +145,18 @@ private Long getFactorsCount(Map<String, List<UUID>> ids, UUID networkUuid, Stri
...
     }
 
-    private Stream<IdentifiableAttributes> getIdentifiablesFromContainer(SensitivityAnalysisRunContext context, EquipmentsContainer filter,
+    private Stream<IdentifiableAttributes> getIdentifiablesFromContainer(SensitivityAnalysisRunContext context, List<EquipmentsContainer> filters,
         List<IdentifiableType> equipmentsTypesAllowed, Reporter reporter) {
-
-        List<IdentifiableAttributes> listIdentAttributes = goGetIdentifiables(filter, context.getNetworkUuid(), context.getVariantId(), reporter);
+        List<String> containerNamesList = filters.stream().map(EquipmentsContainer::getContainerName).toList();
+        String containerNames = Arrays.toString(containerNamesList.toArray());
+        List<IdentifiableAttributes> listIdentAttributes = goGetIdentifiables(filters, context.getNetworkUuid(), context.getVariantId(), reporter);
```
Comment (AAJELLAL) [inline src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java:152]: It's better to avoid abbreviation as names
```suggestion
        List<IdentifiableAttributes> listIdentifiableAttributes = goGetIdentifiables(filters, context.getNetworkUuid(), context.getVariantId(), reporter);
```

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java [hunk 7]
```diff
@@ -161,16 +166,17 @@ private Stream<IdentifiableAttributes> getIdentifiablesFromContainer(Sensitivity
         return listIdentAttributes.stream();
     }
 
-    private Stream<IdentifiableAttributes> getMonitoredIdentifiablesFromContainer(SensitivityAnalysisRunContext context, Network network, EquipmentsContainer filter, List<IdentifiableType> equipmentsTypesAllowed, Reporter reporter) {
-
-        List<IdentifiableAttributes> listIdentAttributes = goGetIdentifiables(filter, context.getNetworkUuid(), context.getVariantId(), reporter);
+    private Stream<IdentifiableAttributes> getMonitoredIdentifiablesFromContainer(SensitivityAnalysisRunContext context, Network network, List<EquipmentsContainer> filters, List<IdentifiableType> equipmentsTypesAllowed, Reporter reporter) {
+        List<String> containerNamesList = filters.stream().map(EquipmentsContainer::getContainerName).toList();
+        String containerNames = Arrays.toString(containerNamesList.toArray());
```
Comment (AAJELLAL) [inline src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInputBuilderService.java:171]: You can refactor this part to avoid code duplication since it's commonly used with **getIdentifiablesFromContainer**
Proposition: 

```
 private String getContainerNames(List<EquipmentsContainer> filters) {
        List<String> containerNamesList = filters.stream().map(EquipmentsContainer::getContainerName).toList();
        return Arrays.toString(containerNamesList.toArray());
    }
```
Then use it here: 
`.withValue("name", getContainerNames(filters))`

File: src/test/java/org/gridsuite/sensitivityanalysis/server/service/FilterServiceTest.java [hunk 8]
```diff
@@ -94,16 +94,16 @@ private String initMockWebServer() throws IOException {
             @Override
             public MockResponse dispatch(RecordedRequest request) {
                 String requestPath = Objects.requireNonNull(request.getPath());
-                if (requestPath.equals(String.format("/v1/filters/%s/export?networkUuid=%s&variantId=%s", LIST_UUID, NETWORK_UUID, VARIANT_ID))) {
+                if (requestPath.equals(String.format("/v1/filters/export?ids=%s&networkUuid=%s&variantId=%s", LIST_UUID, NETWORK_UUID, VARIANT_ID))) {
```
Comment (AAJELLAL) [inline src/test/java/org/gridsuite/sensitivityanalysis/server/service/FilterServiceTest.java:97]: You need to add another test with a list with multiple UUID instead of just one.

**Comment by jamal-khey** [comment]
> I found a problem during a test and reported it but I didn't reproduce it with Jamal yet. Will check back on it monday with Jamal before any approval.

Probleme found , i did a fix and added few tests to detect the issue

**Review by Mathieu-Deharbe — APPROVED** [review]
Crash fixed. Good for me.

**Review by AAJELLAL — APPROVED** [review]
